File: eeg_rpsd.m
function [psdmed, psdvar, psdkrt] = eeg_rpsd(EEG, nfreqs, pct_data)

% clean input cutoff freq
nyquist = floor(EEG.srate / 2);
if ~exist('nfreqs', 'var') || isempty(nfreqs)
    nfreqs = nyquist;
elseif nfreqs > nyquist
    nfreqs = nyquist;
end
if ~exist('pct_data', 'var') || isempty(pct_data)
    pct_data = 100;
end

% setup constants
ncomp = size(EEG.icaweights, 1);
n_points = min(EEG.pnts, EEG.srate);
window = hamming(n_points)';
cutoff = floor(EEG.pnts / n_points) * n_points;
index = bsxfun(@plus, ceil(0:n_points / 2:cutoff - n_points), (1:n_points)');
rng(0)
n_seg = size(index, 2) * EEG.trials;
subset = randperm(n_seg, ceil(n_seg * pct_data / 100)); % need to improve this
rng('shuffle')

try
    %% slightly faster but high memory use
    % calculate windowed spectrums
    temp = reshape(EEG.icaact(:, index, :), [ncomp size(index) .* [1 EEG.trials]]);
    temp = bsxfun(@times, temp(:, :, subset), window);
    temp = fft(temp, n_points, 2);
    temp = abs(temp).^2;
    temp = temp(:, 2:nfreqs + 1, :) * 2 / (EEG.srate*sum(window.^2));
    if nfreqs == nyquist
        temp(:, end, :) = temp(:, end, :) / 2; end

    % calculate outputs
    psdmed = db(median(temp, 3));
    psdvar = db(var(temp, [], 3), 'power');
    psdkrt = db(kurtosis(temp, [], 3)/4);

catch
    %% lower memory but slightly slower
    psdmed = zeros(ncomp, nfreqs);
    psdvar = zeros(ncomp, nfreqs);
    psdkrt = zeros(ncomp, nfreqs);
    for it = 1:ncomp
        % calculate windowed spectrums
        temp = reshape(EEG.icaact(it, index, :), [1 size(index) .* [1 EEG.trials]]);
        temp = bsxfun(@times, temp(:, :, subset), window);
        temp = fft(temp, n_points, 2);
        temp = temp .* conj(temp);
        temp = temp(:, 2:nfreqs + 1, :) * 2 / (EEG.srate*sum(window.^2));
        if nfreqs == nyquist
            temp(:, end, :) = temp(:, end, :) / 2; end

        psdmed(it, :) = db(median(temp, 3));
        psdvar(it, :) = db(var(temp, [], 3), 'power');
        psdkrt(it, :) = db(kurtosis(temp, [], 3)/4);
    end
end
end

File: eeg_rpsd.py
import numpy as np
import math
from scipy.fft import fft

def eeg_rpsd(icaact, icaweights, pnts, srate, nfreqs, trials, pct_data=100, subset = None):
    # Clean input cutoff freq
    nyquist = math.floor(srate / 2)
    if nfreqs > nyquist:
        nfreqs = nyquist
    
    ncomp = len(icaweights)
    n_points = min(pnts, srate)
    window = np.hamming(n_points).reshape(1,-1)[:,:,np.newaxis]

    cutoff = math.floor(pnts / n_points) * n_points
    index = np.ceil(np.arange(0,cutoff - n_points+1,n_points / 2)).astype(np.int64).reshape(1,-1) + np.arange(0,n_points).reshape(-1,1)
    
    n_seg = index.shape[1] * trials
    if subset is None:
        subset = np.random.permutation(n_seg)[:math.ceil(n_seg * pct_data / 100)]
    subset -= 1 # because matlab uses indices starting at 1
    subset = np.squeeze(subset)

    psdmed = np.zeros((ncomp, nfreqs))
    denom = srate * np.sum(np.power(window,2))
    for it in range(ncomp):
        temp = icaact[it, index, :].reshape(1,index.shape[0], n_seg, order='F')
        temp = temp[:,:,subset] * window
        temp = fft(temp, n_points, 1)
        temp = temp * np.conjugate(temp)
        temp = temp[:, 1:nfreqs + 1, :] * 2 / denom
        if nfreqs == nyquist:
            temp[:,-1,:] /= 2
        psdmed[it, :] = 20 * np.real(np.log10(np.median(temp, axis=2)))

    return psdmed
